x/evm/keeper/msg_server.go

```diff
@@ -55,6 +55,285 @@ func validateChainActivated(ctx sdk.Context, n types.Nexus, chain nexus.Chain) e
+
+	queue := keeper.GetContractCallQueue(ctx)
+	txID := req.GetTxID()
+
+	for _, event := range result.Events {
+		key := utils.LowerCaseKey(txID).AppendStr(strconv.FormatUint(event.Info.Index, 10))
+		if queue.Has(key) {
+			return nil, fmt.Errorf("event %s-%d already confirmed", txID, event.Info.Index)
+		}
+
+		switch event.GetEvent().(type) {
+		case *types.VoteConfirmGatewayTxRequest_Vote_Event_ContractCallWithToken:
+			queue.Enqueue(key, &event)
+		default:
+			return nil, fmt.Errorf("unsupported event type %T", event)
+		}
+	}
+
+	return &types.VoteConfirmGatewayTxResponse{}, nil
+}
+
+func (s msgServer) CreateApproveContractCallsFromChain(c context.Context, req *types.CreateApproveContractCallsRequest) (*types.CreateApproveContractCallsResponse, error) {
+	ctx := sdk.UnwrapSDKContext(c)
+
+	chain, ok := s.nexus.GetChain(ctx, req.Chain)
+	if !ok {
+		return nil, fmt.Errorf("%s is not a registered chain", req.Chain)
+	}
+
+	if err := validateChainActivated(ctx, s.nexus, chain); err != nil {
+		return nil, err
+	}
+
+	keeper := s.ForChain(chain.Name)
+	if _, ok := keeper.GetGatewayAddress(ctx); !ok {
+		return nil, fmt.Errorf("axelar gateway address not set")
+	}
+
+	var invalidEvents []types.VoteConfirmGatewayTxRequest_Vote_Event
+
+	queue := keeper.GetContractCallQueue(ctx)
+	var event types.VoteConfirmGatewayTxRequest_Vote_Event
+	for queue.Dequeue(&event) {
+		switch e := event.GetEvent().(type) {
+		case *types.VoteConfirmGatewayTxRequest_Vote_Event_ContractCallWithToken:
+			contractCallWithToken := e.ContractCallWithToken
+
+			token := keeper.GetERC20TokenBySymbol(ctx, contractCallWithToken.Symbol)
+			if !token.Is(types.Confirmed) {
+				s.Logger(ctx).Info(fmt.Sprintf("%s token %s is not confirmed yet", chain.Name, contractCallWithToken.Symbol))
+				invalidEvents = append(invalidEvents, event)
+
+				continue
+			}
+
+			if !s.HasChain(ctx, contractCallWithToken.DestinationChain) {
+				s.Logger(ctx).Info(fmt.Sprintf("destination evm chain %s does not exist", chain.Name))
+				invalidEvents = append(invalidEvents, event)
+
+				continue
+			}
+
+			if token := s.ForChain(contractCallWithToken.DestinationChain).GetERC20TokenByAsset(ctx, token.GetAsset()); !token.Is(types.Confirmed) {
+				s.Logger(ctx).Info(fmt.Sprintf("%s token with asset %s is not confirmed yet", contractCallWithToken.DestinationChain, token.GetAsset()))
+				invalidEvents = append(invalidEvents, event)
+
+				continue
+			}
+
+			if !common.IsHexAddress(contractCallWithToken.ContractAddress) {
+				s.Logger(ctx).Info(fmt.Sprintf("invalid contract address %s for chain %s", contractCallWithToken.ContractAddress, contractCallWithToken.DestinationChain))
+				invalidEvents = append(invalidEvents, event)
+
+				continue
+			}
+
+			destinationChain, ok := s.nexus.GetChain(ctx, contractCallWithToken.DestinationChain)
+			if !ok {
+				return nil, fmt.Errorf("%s is not a registered chain", contractCallWithToken.DestinationChain)
```

Comment:
with this someone can completely break the contract execution. Remember that state is rolled back on error, so if the destination chain is not known, contract commands are stuck